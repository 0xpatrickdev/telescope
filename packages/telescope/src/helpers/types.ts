export const types = `import { BinaryReader, BinaryWriter } from "../../../binary";

/**
 * A type generated by Telescope 1.0.
 */
export interface TelescopeGeneratedType {
  readonly typeUrl: string;
  readonly encode: (
    message:
      | any
      | {
          [k: string]: any;
        },
    writer?: BinaryWriter
  ) => BinaryWriter;
  readonly decode: (input: BinaryReader | Uint8Array, length?: number) => any;
  readonly fromPartial: (object: any) => any;
}

export type GeneratedType = TelescopeGeneratedType;
`;