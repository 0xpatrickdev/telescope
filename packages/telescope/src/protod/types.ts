export interface Repo {
  owner: string;
  repo: string;
}

export interface BufInfo extends Repo {
  git: Repo[];
}

export interface GitInfo extends Repo {
  branch: string;
  protoDir?: string;
  protoPath?: string;
}

export interface DownloadOptions {
  repos?: {
    owner: string;
    repo: string;
    branch?: string;
  }[];
  protoDirMapping?: Record<string, string>;
  outDir?: string;
  ssh?: boolean;
  targets?: string[];
  tempRepoDir?: string;
  deleteTempRepoDir?: boolean;
}

export interface CloneAllOptions {
  repos?: {
    owner: string;
    repo: string;
    branch?: string;
  }[];
  protoDirMapping?: Record<string, string>;
  gitModulesDir: string;
  ssh: boolean;
}

export interface CloneOptions {
  owner: string;
  repo: string;
  branch?: string;
  protoDirMapping?: Record<string, string>;
  gitModulesDir: string;
  ssh: boolean;
  cloned?: Record<string, GitInfo>;
}

export interface ProtoCopyOptions {
  sources: Record<string, GitInfo>;
  targets: string[];
  outDir: string;
}
