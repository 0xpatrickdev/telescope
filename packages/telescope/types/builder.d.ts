import { ProtoStore } from '@osmonauts/proto-parser';
import { TelescopeParseContext } from './build';
import { TelescopeOptions } from '@osmonauts/types';
import { TelescopeInput } from './types';
export declare class TelescopeBuilder {
    store: ProtoStore;
    protoDirs: string[];
    outPath: string;
    options: TelescopeOptions;
    contexts: TelescopeParseContext[];
    files: string[];
    constructor({ protoDirs, outPath, store, options }: TelescopeInput & {
        store?: ProtoStore;
    });
    context(ref: any): TelescopeParseContext;
    build(): void;
}
