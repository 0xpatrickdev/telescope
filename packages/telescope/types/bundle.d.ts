import { ProtoStore } from '@osmonauts/proto-parser';
import { TelescopeBuilder, TelescopeInput } from './index';
import { TelescopeParseContext } from './build';
export declare const getPackages: (store: ProtoStore) => {};
export declare const getPackagesBundled: (store: ProtoStore) => {};
export declare const bundlePackages: (store: ProtoStore, input: TelescopeInput) => {
    bundleFile: string;
    importPaths: any[];
    base: string;
    body: any;
}[];
export declare const bundleRegistries: (telescope: TelescopeBuilder) => {
    package: string;
    contexts: unknown;
}[];
export declare const bundleBaseRegistries: (telescope: TelescopeBuilder) => {
    base: string;
    pkgs: {
        package: string;
        contexts: any;
    }[];
}[];
export declare const parseContextsForRegistry: (contexts: TelescopeParseContext[]) => {
    filename: string;
    imports: {};
    objects: any[];
}[];
export declare const parsePackage: (obj: any, bundleFile: any, importPaths: any, bundleVariables: any) => void;
