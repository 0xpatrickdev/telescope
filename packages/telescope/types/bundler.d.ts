import * as t from '@babel/types';
import { TelescopeParseContext } from './build';
import { TelescopeBuilder } from './builder';
import { ProtoRef } from '@osmonauts/types';
export interface Bundle {
    bundleVariables: {};
    bundleFile: string;
    importPaths: any[];
    base: string;
}
export declare class Bundler {
    builder: TelescopeBuilder;
    contexts: TelescopeParseContext[];
    bundle: Bundle;
    files: string[];
    converters: any[];
    registries: any[];
    constructor(builder: TelescopeBuilder, bundle: Bundle);
    getFreshContext(context: TelescopeParseContext): TelescopeParseContext;
    getLocalFilename(ref: ProtoRef, suffix?: string): string;
    getFilename(localname: string): string;
    writeAst(program: t.Statement[], filename: string): void;
    addToBundle(context: TelescopeParseContext, localname: string): void;
}
