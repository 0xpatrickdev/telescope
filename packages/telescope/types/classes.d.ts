import * as c from '@cosmonauts/ast-gen';
import { Enum, Mutation, EnumConverter, Interface, Field } from '@cosmonauts/ast-gen';
export interface FileStore {
    filename: string;
    code: string;
    program: TSProtoStore;
}
export interface OutFile {
    filename: string;
    ast: any;
}
export interface Import {
    filename: string;
    dep: string;
    import: string;
    type: string;
    node: any;
}
export interface GenericInterface {
    filename: string;
    node: any;
}
export declare class TSProtoStore {
    protoPath: string;
    outPath: string;
    paths: string[];
    files: TSFileStore[];
    enums: Enum[];
    constructor(protoPath: any, outPath: any);
    load(): void;
    write(): void;
    traverse(): void;
    loadFile(filename: string): void;
    getFile(filename: any): TSFileStore;
    getDefinitions(): Interface[];
    getEnum(name: string): Enum;
    addEnum(data: Enum): void;
    getPackages(): any;
    getPackagesBundled(): any;
    bundlePackages(): void;
    bundleRootPackage(): void;
    protoPathToOutPath(filename: any): string;
    writeFile(ast: any, filename: any): void;
    findDependency(name: any): {
        type: string;
        name: any;
        loc: string;
    };
    findFilesOfPackage(protoPackage: string): TSFileStore[];
    findInterfaceFile(type: string): TSFileStore;
    findInterfaceFiles(type: string): TSFileStore[];
    print(): void;
}
export declare class TSFileStore implements FileStore {
    filename: string;
    protoPath: string;
    outPath: string;
    protoPackage: string;
    code: string;
    ast: any;
    program: TSProtoStore;
    interfaces: GenericInterface[];
    definitions: Interface[];
    imports: Import[];
    exports: any[];
    queries: any[];
    mutations: Mutation[];
    enums: Enum[];
    visitor: any;
    mutator: any;
    _traversed: boolean;
    importStmts: any[];
    outFiles: OutFile[];
    constructor(filename: any, code: any, program: any);
    traverse(): void;
    isQuery(): boolean;
    isTransaction(): boolean;
    setPackage(value: any): void;
    addGenericImport(data: any): void;
    addImport(data: any): void;
    addGenericInterface(node: any): void;
    addInterfaceField(type: string, el: Field): void;
    getInterface(type: string): Interface;
    addEnum(el: Enum): void;
    getEnum(name: string): Enum;
    hasEnum(name: string): boolean;
    addEnumFunction(key: string, data: EnumConverter): void;
    writeFile(ast: any, filename: any): void;
    getSiblingFileName(name: any): string;
    write(): void;
    generateFiles(): void;
    print(): void;
    toJSON(): {
        filename: string;
        protoPackage: string;
        mutations: c.Mutation[];
    };
}
