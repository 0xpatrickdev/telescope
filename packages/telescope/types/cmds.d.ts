export namespace Commands {
    export { _generate as generate };
    export { _transpile as transpile };
}
import _generate from "./commands/generate";
import _transpile from "./commands/transpile";
