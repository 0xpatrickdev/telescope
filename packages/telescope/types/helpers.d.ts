export declare const aminoHelperCode: () => string;
