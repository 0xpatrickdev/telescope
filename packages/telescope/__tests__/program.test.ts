import generate from '@babel/generator';
import * as t from '@babel/types';
import { tmpdir } from 'os';
import { join } from 'path';
import {
    TSProtoStore
} from '../src'

const protoPath = __dirname + '/../__fixtures__/chain1'
const outPath = join(tmpdir(), 'chain1');

const program = new TSProtoStore(protoPath, outPath);

it('packages', async () => {
    const superfluid = program.findFilesOfPackage('osmosis.superfluid');
    const gamm = program.findFilesOfPackage('osmosis.gamm.v1beta1');
    expect(superfluid.map(s => (s.toJSON()))).toMatchSnapshot();
    expect(gamm.map(s => (s.toJSON()))).toMatchSnapshot();
})

it('messages', async () => {
    const file = program.findInterfaceFile('MsgCreateBalancerPool');
    expect(file.getInterface('MsgCreateBalancerPool')).toBeTruthy();
});

it('interfaces', async () => {
    const file = program.findInterfaceFile('MsgJoinPool');
    expect(file.getInterface('MsgJoinPool')).toBeTruthy();
    const i = file.getInterface('MsgJoinPool').fields.reduce((m, v) => {
        m[v.name] = v.type;
        return m;
    }, {});
    expect(i).toMatchSnapshot();
});

it('files', async () => {
    const file = program.findInterfaceFile('MsgJoinPool');
    file.write();
});

it('VoteOption', async () => {
    const files = program.findInterfaceFiles('MsgVote');
    expect(files.length).toBe(3);
});

it('enums', async () => {
    const enum1 = program.getEnum('Exec');
    expect(enum1).toBeTruthy();
    const filename = enum1.filename;
    const file = program.getFile(filename);
    const enum2 = file.getEnum('Exec');
    expect(enum1).toEqual(enum2);
});

it('dependencies', async () => {
    const a = program.findDependency('Exec');
    const b = program.findDependency('VoteOption');
    const c = program.findDependency('MsgJoinPool');
    a.loc = a.loc.replace(program.protoPath, '')
    b.loc = b.loc.replace(program.protoPath, '')
    c.loc = c.loc.replace(program.protoPath, '')
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(c).toMatchSnapshot();
});

it('non-TX deps', async () => {
    const a = program.findDependency('EventGrant');
    a.loc = a.loc.replace(program.protoPath, '')
    expect(a).toMatchSnapshot();
});

it('Nested Props', async () => {
    const a = program.findDependency('Description');
    const b = program.findDependency('CommissionRates');
    a.loc = a.loc.replace(program.protoPath, '')
    b.loc = b.loc.replace(program.protoPath, '')
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    const staking = program.findInterfaceFile('CommissionRates');
    const comm = staking.getInterface('CommissionRates');
    const desc = staking.getInterface('Description');
    expect(desc).toBeTruthy();
    expect(comm).toBeTruthy();

    // // console.log(comm, desc)
    // desc.fields.forEach(field => {
    //     const val = generate(field.node).code;

    //     // console.log(field.name, val);
    // })


});

// TODO
// can we ask the program, 
// hey, where can i get omitDefault?
// and it returns the aminos?
// feels like maybe we should use TSFile for every file?
// but maybe we can generate new files in the process of parsing files.

it('AminoHelpers', async () => {
    const a = program.findDependency('AminoHeight');
    a.loc = a.loc.replace(program.protoPath, '')
    expect(a).toMatchSnapshot();
});

it('program', async () => {
    program.write();
});
