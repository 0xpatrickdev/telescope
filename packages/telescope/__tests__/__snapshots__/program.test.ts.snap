// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AminoHelpers 1`] = `
Object {
  "loc": "/amino.helpers.ts",
  "name": "AminoHeight",
  "type": "Interface",
}
`;

exports[`Nested Props 1`] = `
Object {
  "loc": "/cosmos/staking/v1beta1/staking.ts",
  "name": "Description",
  "type": "Interface",
}
`;

exports[`Nested Props 2`] = `
Object {
  "loc": "/cosmos/staking/v1beta1/staking.ts",
  "name": "CommissionRates",
  "type": "Interface",
}
`;

exports[`dependencies 1`] = `
Object {
  "loc": "/cosmos/group/v1/tx.ts",
  "name": "Exec",
  "type": "Enum",
}
`;

exports[`dependencies 2`] = `
Object {
  "loc": "/cosmos/gov/v1/gov.ts",
  "name": "VoteOption",
  "type": "Enum",
}
`;

exports[`dependencies 3`] = `
Object {
  "loc": "/osmosis/gamm/v1beta1/tx.ts",
  "name": "MsgJoinPool",
  "type": "Interface",
}
`;

exports[`interfaces 1`] = `
Object {
  "poolId": "Long",
  "sender": "string",
  "shareOutAmount": "string",
  "tokenInMaxs": "Coin[]",
}
`;

exports[`non-TX deps 1`] = `
Object {
  "loc": "/cosmos/authz/v1beta1/event.ts",
  "name": "EventGrant",
  "type": "Interface",
}
`;

exports[`packages 1`] = `
Array [
  Object {
    "filename": "/osmosis/superfluid/genesis.ts",
    "mutations": Array [],
    "protoPackage": "osmosis.superfluid",
  },
  Object {
    "filename": "/osmosis/superfluid/params.ts",
    "mutations": Array [],
    "protoPackage": "osmosis.superfluid",
  },
  Object {
    "filename": "/osmosis/superfluid/query.ts",
    "mutations": Array [],
    "protoPackage": "osmosis.superfluid",
  },
  Object {
    "filename": "/osmosis/superfluid/superfluid.ts",
    "mutations": Array [],
    "protoPackage": "osmosis.superfluid",
  },
  Object {
    "filename": "/osmosis/superfluid/tx.ts",
    "mutations": Array [
      Object {
        "TypeName": "MsgSuperfluidDelegate",
        "methodName": "superfluidDelegate",
        "typeUrl": "/osmosis.superfluid.MsgSuperfluidDelegate",
      },
      Object {
        "TypeName": "MsgSuperfluidUndelegate",
        "methodName": "superfluidUndelegate",
        "typeUrl": "/osmosis.superfluid.MsgSuperfluidUndelegate",
      },
      Object {
        "TypeName": "MsgSuperfluidUnbondLock",
        "methodName": "superfluidUnbondLock",
        "typeUrl": "/osmosis.superfluid.MsgSuperfluidUnbondLock",
      },
      Object {
        "TypeName": "MsgLockAndSuperfluidDelegate",
        "methodName": "lockAndSuperfluidDelegate",
        "typeUrl": "/osmosis.superfluid.MsgLockAndSuperfluidDelegate",
      },
    ],
    "protoPackage": "osmosis.superfluid",
  },
]
`;

exports[`packages 2`] = `
Array [
  Object {
    "filename": "/osmosis/gamm/pool-models/balancer/balancerPool.ts",
    "mutations": Array [],
    "protoPackage": "osmosis.gamm.v1beta1",
  },
  Object {
    "filename": "/osmosis/gamm/pool-models/balancer/tx.ts",
    "mutations": Array [],
    "protoPackage": "osmosis.gamm.v1beta1",
  },
  Object {
    "filename": "/osmosis/gamm/v1beta1/genesis.ts",
    "mutations": Array [],
    "protoPackage": "osmosis.gamm.v1beta1",
  },
  Object {
    "filename": "/osmosis/gamm/v1beta1/query.ts",
    "mutations": Array [],
    "protoPackage": "osmosis.gamm.v1beta1",
  },
  Object {
    "filename": "/osmosis/gamm/v1beta1/tx.ts",
    "mutations": Array [
      Object {
        "TypeName": "MsgJoinPool",
        "methodName": "joinPool",
        "typeUrl": "/osmosis.gamm.v1beta1.MsgJoinPool",
      },
      Object {
        "TypeName": "MsgExitPool",
        "methodName": "exitPool",
        "typeUrl": "/osmosis.gamm.v1beta1.MsgExitPool",
      },
      Object {
        "TypeName": "MsgSwapExactAmountIn",
        "methodName": "swapExactAmountIn",
        "typeUrl": "/osmosis.gamm.v1beta1.MsgSwapExactAmountIn",
      },
      Object {
        "TypeName": "MsgSwapExactAmountOut",
        "methodName": "swapExactAmountOut",
        "typeUrl": "/osmosis.gamm.v1beta1.MsgSwapExactAmountOut",
      },
      Object {
        "TypeName": "MsgJoinSwapExternAmountIn",
        "methodName": "joinSwapExternAmountIn",
        "typeUrl": "/osmosis.gamm.v1beta1.MsgJoinSwapExternAmountIn",
      },
      Object {
        "TypeName": "MsgJoinSwapShareAmountOut",
        "methodName": "joinSwapShareAmountOut",
        "typeUrl": "/osmosis.gamm.v1beta1.MsgJoinSwapShareAmountOut",
      },
      Object {
        "TypeName": "MsgExitSwapExternAmountOut",
        "methodName": "exitSwapExternAmountOut",
        "typeUrl": "/osmosis.gamm.v1beta1.MsgExitSwapExternAmountOut",
      },
      Object {
        "TypeName": "MsgExitSwapShareAmountIn",
        "methodName": "exitSwapShareAmountIn",
        "typeUrl": "/osmosis.gamm.v1beta1.MsgExitSwapShareAmountIn",
      },
    ],
    "protoPackage": "osmosis.gamm.v1beta1",
  },
]
`;
